### PR TITLE
[Pytorch] Bugfix in te fusion ce implementation

### PR DESCRIPTION
**Pull request by BestJuly**
# Description

Fix the issue in [Megatron-LM#1627](https://github.com/NVIDIA/Megatron-LM/issues/1627).    
Add related unit test that capture this.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/pytorch/test_parallel_cross_entropy.py [hunk 1]
```diff
@@ -61,19 +62,33 @@ def one_iteration_test(
...
             ref_loss.backward()
 
-        test_loss = torch.flatten(test_loss) if not reduce_loss else test_loss
+        # Compare losses
+        if use_sum_backward:
+            # When using sum().backward(), compare the summed losses
+            torch.testing.assert_close(test_loss.sum(), ref_loss.sum(), check_dtype=False)
+        else:
+            test_loss = torch.flatten(test_loss) if not reduce_loss else test_loss
+            torch.testing.assert_close(test_loss, ref_loss, check_dtype=False)
```
Comment (timmoon10) [inline tests/pytorch/test_parallel_cross_entropy.py]: We still want to check that each entry of the loss output is correct, even if we're going to sum afterward:
```suggestion
        test_loss = torch.flatten(test_loss) if not reduce_loss else test_loss
```

File: tests/pytorch/test_parallel_cross_entropy.py [hunk 2]
```diff
@@ -51,6 +51,7 @@ def one_iteration_test(
         label_smoothing: float,
         reduce_loss: bool,
         ignore_idx: bool = False,
+        use_sum_backward: bool = False,
```
Comment (timmoon10) [inline tests/pytorch/test_parallel_cross_entropy.py]: Why not check  `not reduce_loss`? We don't check the backward pass when `reduce_loss=False` and `use_sum_backward=False`, which seems like a hole in our testing.
```suggestion
```
This also means we don't need to add a new test case.

**Review by timmoon10 — COMMENTED** [review]
Overall makes sense, but we should change the tests so that every test case checks the backward pass. We missed this bug because we only tested the backward pass with `reduce_loss=True`.

File: tests/pytorch/test_parallel_cross_entropy.py (hunk 2 above)
Comment (BestJuly) [reply tests/pytorch/test_parallel_cross_entropy.py]: That is a good suggestion. I have modified the test code accordingly.

File: tests/pytorch/test_parallel_cross_entropy.py [hunk 3]
```diff
@@ -134,3 +138,19 @@ def test_ignore_idx(self):
...
+    #     """
+    #     # Set up infrastructure for non-reduced loss (reduction='none')
+    #     self.generate_infra(False, 0)  # reduce_loss=False, label_smoothing=0
+    #     self.one_iteration_test(
+    #         dtype=torch.float32,
+    #         swap_dim=False,
+    #         label_smoothing=0,
+    #         reduce_loss=False,
+    #         ignore_idx=False,
+    #     )
```
Comment (timmoon10) [inline tests/pytorch/test_parallel_cross_entropy.py]:
```suggestion
```

**Review by timmoon10 — APPROVED** [review]
LGTM, pending CI.

**Review by ksivaman — APPROVED** [review]
